File: turtlebot.py
import os
import requests
import turtle


SPARK_URL = 'https://api.spark.io/v1/devices/{device}/{action}/'


def send(action, *args):
    device = os.environ['TURTLEBOT_SPARK_DEVICE']
    access_token = os.environ['TURTLEBOT_SPARK_TOKEN']
    method = 'POST'

    url = SPARK_URL.format(device=device, action=action)
    data = {
        'access_token': access_token,
    }
    if args:
        data['args'] = str(args[0])

    print(method, url, data)
    session = requests.session()
    response = session.request(method, url, data=data)
    print(response)
    print(response.text)

def forward(distance):
    print('forward', distance)
    turtle.forward(distance)
    send('move', distance)

def backward(distance):
    print('backward', distance)
    turtle.backward(distance)
    send('move', -distance)

def left(angle):
    print('left', angle)
    turtle.left(angle)
    send('rotate', -angle)

def right(angle):
    print('right', angle)
    turtle.right(angle)
    send('rotate', -angle)
    
def penup():
    print('penup')
    turtle.penup()
    send('penmove', 1)

def pendown():
    print('pendown')
    turtle.pendown()
    send('penmove', -1)


def almost_rectangle_test():
    forward(50)

    left(90)
    forward(50)

    penup()
    right(90)
    backward(50)

    pendown()
    left(90)
    backward(50)

File: web/js/turtlebot.js
var turtlebot = {
    penIsDown: false,

    sparkConfigured: function() {
        return typeof spark != 'undefined';
    },

    execute: function (command, args) {
        console.log('execute ' + command + ' ' + args);
        if (this.sparkConfigured()) {
            $.post('https://api.spark.io/v1/devices/' + spark.deviceId + '/' + command, {
                access_token: spark.accessToken,
                args: args
            });
        } else {
            console.error('Spark not configured. See spark.js.template for details')
        }
    },

    forward: function (distance) {
        console.log('forward');
        this.execute('move', distance);
    },

    backward: function (distance) {
        console.log('backward');
        this.execute('move', -distance);
    },

    left: function (angle) {
        console.log('left');
        this.execute('rotate', -angle);
    },

    right: function (angle) {
        console.log('right');
        this.execute('rotate', angle);
    },

    togglePen: function () {
        this.penIsDown ? this.penUp() : this.penDown();
    },

    penUp: function () {
        console.log('pen up');
        this.execute('penmove', 1);
        this.penIsDown = false;
    },

    penDown: function () {
        console.log('pen down');
        this.execute('penmove', -1);
        this.penIsDown = true;
    }
};

function numericFieldValue(fieldId) {
    var textValue = $('#' + fieldId).val();
    return parseInt(textValue);
}

$('#forward').click(function () {
    var distance = numericFieldValue('forwardDistance');
    turtlebot.forward(distance);
});

$('#backward').click(function () {
    var distance = numericFieldValue('backwardDistance');
    turtlebot.backward(distance);
});

$('#left').click(function () {
    var angle = numericFieldValue('leftAngle');
    turtlebot.left(angle);
});

$('#right').click(function () {
    var angle = numericFieldValue('rightAngle');
    turtlebot.right(angle);
});

$('#pen').click(function () {
    turtlebot.togglePen();
});

$(function () {
    if (!turtlebot.sparkConfigured()) {
        $('#sparkNotConfiguredMessage').show();
    }
});
